druid/src/text/editor.rs

```diff
@@ -0,0 +1,311 @@
+    /// Must be called in WidgetAdded
+    pub fn rebuild_if_needed(&mut self, factory: &mut PietText, env: &Env) {
+        self.layout.rebuild_if_needed(factory, env);
+    }
+
+    /// Perform an [`EditAction`](enum.EditAction.html).
+    pub fn do_edit(&mut self, edit: EditAction, data: &mut T) {
+        if self.data_is_stale(data) {
+            log::warn!("editor data changed externally, skipping event {:?}", &edit);
+            return;
+        }
+        match edit {
+            EditAction::Insert(chars) | EditAction::Paste(chars) => self.insert(data, &chars),
+            EditAction::Backspace => self.delete_backward(data),
+            EditAction::Delete => self.delete_forward(data),
+            EditAction::JumpDelete(mvmt) | EditAction::JumpBackspace(mvmt) => {
+                let to_delete = if self.selection.is_caret() {
+                    movement(mvmt, self.selection, data, true)
+                } else {
+                    self.selection
+                };
+                data.edit(to_delete.range(), "");
+                self.selection = Selection::caret(to_delete.min());
+            }
+            EditAction::Move(mvmt) => self.selection = movement(mvmt, self.selection, data, false),
+            EditAction::ModifySelection(mvmt) => {
+                self.selection = movement(mvmt, self.selection, data, true)
+            }
+            EditAction::Click(action) => {
+                if action.mods.shift() {
+                    self.selection.end = action.column;
+                } else {
+                    self.selection = Selection::caret(action.column);
+                }
+            }
+            EditAction::Drag(action) => self.selection.end = action.column,
+            EditAction::SelectAll => self.selection = Selection::new(0, data.len()),
+        }
+    }
+
+    /// Draw this editor at the provided point.
+    pub fn draw(&self, ctx: &mut PaintCtx, point: impl Into<Point>) {
+        self.layout.draw(ctx, point)
+    }
+
+    /// Returns `true` if the data passed here has been changed externally,
+    /// which means things like our selection state may be out of sync.
+    ///
+    /// This would only happen in the unlikely case that somebody else has mutated
+    /// the data before us while handling an event; if this is the case we ignore
+    /// the event, and our data will be updated in `update`.
+    fn data_is_stale(&self, data: &T) -> bool {
+        self.layout.text().map(|t| !t.same(data)).unwrap_or(true)
+    }
+
+    fn insert(&mut self, data: &mut T, text: &str) {
+        // if we aren't multiline, we insert only up to the first newline
+        let text = if self.multiline {
+            text
+        } else {
+            text.split('\n').next().unwrap_or("")
+        };
+        let sel = self.selection.range();
+        data.edit(sel, text);
+        self.selection = Selection::caret(self.selection.min() + text.len());
+    }
+
+    /// Delete to previous grapheme if in caret mode.
```

Comment:
This comment is imprecise: the backspace logic sometimes uses backspace granularity, but other times uses finer granularity to match expected behavior (individual vowels and diacritics in Indic scripts, harakat in Arabic, etc).